// @ngx-i18nsupport/ngx-i18nsupport/esm2015/autotranslate/auto-translate-service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { format } from 'util';
import * as request from 'request';
import { Observable } from 'rxjs';
import { of, forkJoin, throwError } from 'rxjs';
import { map } from 'rxjs/operators';
/**
 * Types form google translate api.
 * @record
 */
function GetSupportedLanguagesRequest() { }
if (false) {
    /** @type {?} */
    GetSupportedLanguagesRequest.prototype.target;
}
/**
 * @record
 */
function LanguagesResource() { }
if (false) {
    /** @type {?} */
    LanguagesResource.prototype.language;
    /** @type {?} */
    LanguagesResource.prototype.name;
}
/**
 * @record
 */
function LanguagesListResponse() { }
if (false) {
    /** @type {?} */
    LanguagesListResponse.prototype.languages;
}
/**
 * @record
 */
function TranslateTextRequest() { }
if (false) {
    /** @type {?} */
    TranslateTextRequest.prototype.q;
    /** @type {?} */
    TranslateTextRequest.prototype.target;
    /** @type {?} */
    TranslateTextRequest.prototype.source;
    /** @type {?|undefined} */
    TranslateTextRequest.prototype.format;
    /** @type {?|undefined} */
    TranslateTextRequest.prototype.model;
}
/**
 * @record
 */
function TranslationsResource() { }
if (false) {
    /** @type {?|undefined} */
    TranslationsResource.prototype.detectedSourceLanguage;
    /** @type {?|undefined} */
    TranslationsResource.prototype.model;
    /** @type {?} */
    TranslationsResource.prototype.translatedText;
}
/**
 * @record
 */
function TranslationsListResponse() { }
if (false) {
    /** @type {?} */
    TranslationsListResponse.prototype.translations;
}
/**
 * @record
 */
function InternalRequestResponse() { }
if (false) {
    /** @type {?} */
    InternalRequestResponse.prototype.response;
    /** @type {?} */
    InternalRequestResponse.prototype.body;
}
/** @type {?} */
const MAX_SEGMENTS = 128;
export class AutoTranslateService {
    /**
     * Strip region code and convert to lower
     * @param {?} lang lang
     * @return {?} lang without region code and in lower case.
     */
    static stripRegioncode(lang) {
        /** @type {?} */
        const langLower = lang.toLowerCase();
        for (let i = 0; i < langLower.length; i++) {
            /** @type {?} */
            const c = langLower.charAt(i);
            if (c < 'a' || c > 'z') {
                return langLower.substring(0, i);
            }
        }
        return langLower;
    }
    /**
     * @param {?} apiKey
     */
    constructor(apiKey) {
        this._request = request;
        this._apiKey = apiKey;
        this._rootUrl = 'https://translation.googleapis.com/';
    }
    /**
     * Change API key (just for tests).
     * @param {?} apikey apikey
     * @return {?}
     */
    setApiKey(apikey) {
        this._apiKey = apikey;
    }
    /**
     * Translate an array of messages at once.
     * @param {?} messages the messages to be translated
     * @param {?} from source language code
     * @param {?} to target language code
     * @return {?} Observable with translated messages or error
     */
    translateMultipleStrings(messages, from, to) {
        // empty array needs no translation and always works ... (#78)
        if (messages.length === 0) {
            return of([]);
        }
        if (!this._apiKey) {
            return throwError('cannot autotranslate: no api key');
        }
        if (!from || !to) {
            return throwError('cannot autotranslate: source and target language must be set');
        }
        from = AutoTranslateService.stripRegioncode(from);
        to = AutoTranslateService.stripRegioncode(to);
        /** @type {?} */
        const allRequests = this.splitMessagesToGoogleLimit(messages).map((/**
         * @param {?} partialMessages
         * @return {?}
         */
        (partialMessages) => {
            return this.limitedTranslateMultipleStrings(partialMessages, from, to);
        }));
        return forkJoin(allRequests).pipe(map((/**
         * @param {?} allTranslations
         * @return {?}
         */
        (allTranslations) => {
            /** @type {?} */
            let all = [];
            for (let i = 0; i < allTranslations.length; i++) {
                all = all.concat(allTranslations[i]);
            }
            return all;
        })));
    }
    /**
     * @private
     * @param {?} messages
     * @return {?}
     */
    splitMessagesToGoogleLimit(messages) {
        if (messages.length <= MAX_SEGMENTS) {
            return [messages];
        }
        /** @type {?} */
        const result = [];
        /** @type {?} */
        let currentPackage = [];
        /** @type {?} */
        let packageSize = 0;
        for (let i = 0; i < messages.length; i++) {
            currentPackage.push(messages[i]);
            packageSize++;
            if (packageSize >= MAX_SEGMENTS) {
                result.push(currentPackage);
                currentPackage = [];
                packageSize = 0;
            }
        }
        if (currentPackage.length > 0) {
            result.push(currentPackage);
        }
        return result;
    }
    /**
     * Return translation request, but messages must be limited to google limits.
     * Not more that 128 single messages.
     * @private
     * @param {?} messages messages
     * @param {?} from from
     * @param {?} to to
     * @return {?} the translated strings
     */
    limitedTranslateMultipleStrings(messages, from, to) {
        /** @type {?} */
        const realUrl = this._rootUrl + 'language/translate/v2' + '?key=' + this._apiKey;
        /** @type {?} */
        const translateRequest = {
            q: messages,
            target: to,
            source: from,
        };
        /** @type {?} */
        const options = {
            url: realUrl,
            body: translateRequest,
            json: true,
        };
        return this.post(realUrl, options).pipe(map((/**
         * @param {?} data
         * @return {?}
         */
        (data) => {
            /** @type {?} */
            const body = data.body;
            if (!body) {
                throw new Error('no result received');
            }
            if (body.error) {
                if (body.error.code === 400) {
                    if (body.error.message === 'Invalid Value') {
                        throw new Error(format('Translation from "%s" to "%s" not supported', from, to));
                    }
                    throw new Error(format('Invalid request: %s', body.error.message));
                }
                else {
                    throw new Error(format('Error %s: %s', body.error.code, body.error.message));
                }
            }
            /** @type {?} */
            const result = body.data;
            return result.translations.map((/**
             * @param {?} translation
             * @return {?}
             */
            (translation) => {
                return translation.translatedText;
            }));
        })));
    }
    /**
     * Function to do a POST HTTP request
     *
     * @param {?} uri uri
     * @param {?=} options options
     *
     * @return {?} response
     */
    post(uri, options) {
        return (/** @type {?} */ (this._call.apply(this, [].concat('post', (/** @type {?} */ (uri)), (/** @type {?} */ (Object.assign({}, options || {})))))));
    }
    /**
     * Function to do a HTTP request for given method
     *
     * @private
     * @param {?} method method
     * @param {?} uri uri
     * @param {?=} options options
     *
     * @return {?} response
     *
     */
    _call(method, uri, options) {
        return (/** @type {?} */ (Observable.create((/**
         * @param {?} observer
         * @return {?}
         */
        (observer) => {
            // build params array
            /** @type {?} */
            const params = [].concat((/** @type {?} */ (uri)), (/** @type {?} */ (Object.assign({}, options || {}))), (/**
             * @template RequestCallback
             * @param {?} error
             * @param {?} response
             * @param {?} body
             * @return {?}
             */
            (error, response, body) => {
                if (error) {
                    return observer.error(error);
                }
                observer.next((/** @type {?} */ (Object.assign({}, {
                    response: (/** @type {?} */ (response)),
                    body: (/** @type {?} */ (body))
                }))));
                observer.complete();
            }));
            // _call request method
            try {
                this._request[(/** @type {?} */ (method))].apply((/** @type {?} */ (this._request)), params);
            }
            catch (error) {
                observer.error(error);
            }
        }))));
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    AutoTranslateService.prototype._request;
    /** @type {?} */
    AutoTranslateService.prototype._rootUrl;
    /** @type {?} */
    AutoTranslateService.prototype._apiKey;
}
