// ng-mocks/dist/lib/mock-module/test-fixtures.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var common_1 = require("@angular/common");
var core_1 = require("@angular/core");
var router_1 = require("@angular/router");
var ExampleDirective = (function () {
    function ExampleDirective(node) {
        this.node = node;
    }
    ExampleDirective.prototype.ngOnInit = function () {
        this.node.nativeElement.innerText = 'ExampleDirective';
    };
    ExampleDirective = __decorate([
        core_1.Directive({ selector: '[example-directive]' }),
        __metadata("design:paramtypes", [core_1.ElementRef])
    ], ExampleDirective);
    return ExampleDirective;
}());
exports.ExampleDirective = ExampleDirective;
var ExamplePipe = (function () {
    function ExamplePipe() {
        this.transform = function (text) { return "Example: " + text; };
    }
    ExamplePipe = __decorate([
        core_1.Pipe({ name: 'examplePipe' })
    ], ExamplePipe);
    return ExamplePipe;
}());
exports.ExamplePipe = ExamplePipe;
var ExampleService = (function () {
    function ExampleService() {
        this.get = function (id) { return "Got: " + id; };
    }
    ExampleService = __decorate([
        core_1.Injectable()
    ], ExampleService);
    return ExampleService;
}());
exports.ExampleService = ExampleService;
var ExamplePrivateComponent = (function () {
    function ExamplePrivateComponent() {
    }
    ExamplePrivateComponent = __decorate([
        core_1.Component({
            selector: 'example-private-component',
            template: '<span>Private thing</span>'
        })
    ], ExamplePrivateComponent);
    return ExamplePrivateComponent;
}());
exports.ExamplePrivateComponent = ExamplePrivateComponent;
var ExampleComponent = (function () {
    function ExampleComponent() {
    }
    ExampleComponent = __decorate([
        core_1.Component({
            selector: 'example-component',
            template: '<span>My Example</span>'
        })
    ], ExampleComponent);
    return ExampleComponent;
}());
exports.ExampleComponent = ExampleComponent;
var ExampleConsumerComponent = (function () {
    function ExampleConsumerComponent() {
    }
    ExampleConsumerComponent = __decorate([
        core_1.Component({
            selector: 'example-consumer-component',
            template: '<example-component></example-component>'
        })
    ], ExampleConsumerComponent);
    return ExampleConsumerComponent;
}());
exports.ExampleConsumerComponent = ExampleConsumerComponent;
var ChildModule = (function () {
    function ChildModule() {
    }
    ChildModule = __decorate([
        core_1.NgModule({
            declarations: [ExamplePrivateComponent, ExampleComponent, ExamplePipe, ExampleDirective],
            exports: [ExampleComponent, ExamplePipe, ExampleDirective],
            imports: [common_1.CommonModule],
            providers: [ExampleService]
        })
    ], ChildModule);
    return ChildModule;
}());
exports.ChildModule = ChildModule;
var ParentModule = (function () {
    function ParentModule() {
    }
    ParentModule = __decorate([
        core_1.NgModule({
            imports: [ChildModule],
        })
    ], ParentModule);
    return ParentModule;
}());
exports.ParentModule = ParentModule;
var SameImports1Module = (function () {
    function SameImports1Module() {
    }
    SameImports1Module = __decorate([
        core_1.NgModule({
            imports: [ParentModule],
        })
    ], SameImports1Module);
    return SameImports1Module;
}());
exports.SameImports1Module = SameImports1Module;
var SameImports2Module = (function () {
    function SameImports2Module() {
    }
    SameImports2Module = __decorate([
        core_1.NgModule({
            imports: [ParentModule],
        })
    ], SameImports2Module);
    return SameImports2Module;
}());
exports.SameImports2Module = SameImports2Module;
var LogicNestedModule = (function () {
    function LogicNestedModule() {
    }
    LogicNestedModule = __decorate([
        core_1.NgModule({
            imports: [ChildModule],
        })
    ], LogicNestedModule);
    return LogicNestedModule;
}());
exports.LogicNestedModule = LogicNestedModule;
var LogicRootModule = (function () {
    function LogicRootModule() {
    }
    LogicRootModule = __decorate([
        core_1.NgModule({
            imports: [ChildModule, LogicNestedModule],
        })
    ], LogicRootModule);
    return LogicRootModule;
}());
exports.LogicRootModule = LogicRootModule;
var AppRoutingModule = (function () {
    function AppRoutingModule() {
    }
    AppRoutingModule = __decorate([
        core_1.NgModule({
            imports: [router_1.RouterModule.forRoot([
                    {
                        component: ExampleComponent,
                        path: '',
                    },
                ])],
        })
    ], AppRoutingModule);
    return AppRoutingModule;
}());
exports.AppRoutingModule = AppRoutingModule;
var RealModuleWithProvidersModule = (function () {
    function RealModuleWithProvidersModule() {
    }
    RealModuleWithProvidersModule = __decorate([
        core_1.NgModule({
            imports: [
                common_1.CommonModule,
            ],
        })
    ], RealModuleWithProvidersModule);
    return RealModuleWithProvidersModule;
}());
var ModuleProvider = (function () {
    function ModuleProvider() {
    }
    ModuleProvider.withFlag = function (flag) {
        return {
            ngModule: RealModuleWithProvidersModule,
            providers: [
                {
                    provide: 'MODULE_FLAG',
                    useValue: flag,
                },
            ],
        };
    };
    return ModuleProvider;
}());
var ModuleWithProvidersModule = (function () {
    function ModuleWithProvidersModule() {
    }
    ModuleWithProvidersModule = __decorate([
        core_1.NgModule({
            imports: [
                ModuleProvider.withFlag(false),
            ],
        })
    ], ModuleWithProvidersModule);
    return ModuleWithProvidersModule;
}());
exports.ModuleWithProvidersModule = ModuleWithProvidersModule;
