// @ngx-i18nsupport/ngx-i18nsupport/esm5/xliffmerge/ngx-translate-extraction-pattern.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * Helper class to parse ngx translate extraction pattern
 * and to decide wether a given message matches the pattern.
 */
var /**
 * Helper class to parse ngx translate extraction pattern
 * and to decide wether a given message matches the pattern.
 */
NgxTranslateExtractionPattern = /** @class */ (function () {
    /**
     * Construct the pattern from given description string
     * @param extractionPatternString extractionPatternString
     * @throws an error, if there is a syntax error
     */
    function NgxTranslateExtractionPattern(extractionPatternString) {
        this.extractionPatternString = extractionPatternString;
        /** @type {?} */
        var parts = extractionPatternString.split('|');
        this._matchExplicitId = false;
        this._descriptionPatterns = [];
        for (var i = 0; i < parts.length; i++) {
            /** @type {?} */
            var part = parts[i];
            if (part === '@@') {
                if (this._matchExplicitId) {
                    throw new Error('extraction pattern must not contain @@ twice');
                }
                this._matchExplicitId = true;
            }
            else {
                /** @type {?} */
                var errorString = this.checkValidDescriptionPattern(part);
                if (errorString) {
                    throw new Error(errorString);
                }
                this._descriptionPatterns.push(part);
            }
        }
    }
    /**
     * Check, wether an explicitly set id matches the pattern.
     * @param id id
     * @return wether an explicitly set id matches the pattern.
     */
    /**
     * Check, wether an explicitly set id matches the pattern.
     * @param {?} id id
     * @return {?} wether an explicitly set id matches the pattern.
     */
    NgxTranslateExtractionPattern.prototype.isExplicitIdMatched = /**
     * Check, wether an explicitly set id matches the pattern.
     * @param {?} id id
     * @return {?} wether an explicitly set id matches the pattern.
     */
    function (id) {
        return id && this._matchExplicitId;
    };
    /**
     * Check, wether a given description matches the pattern.
     * @param description description
     * @return wether a given description matches the pattern.
     */
    /**
     * Check, wether a given description matches the pattern.
     * @param {?} description description
     * @return {?} wether a given description matches the pattern.
     */
    NgxTranslateExtractionPattern.prototype.isDescriptionMatched = /**
     * Check, wether a given description matches the pattern.
     * @param {?} description description
     * @return {?} wether a given description matches the pattern.
     */
    function (description) {
        return this._descriptionPatterns.indexOf(description) >= 0;
    };
    /**
     * @private
     * @param {?} descriptionPattern
     * @return {?}
     */
    NgxTranslateExtractionPattern.prototype.checkValidDescriptionPattern = /**
     * @private
     * @param {?} descriptionPattern
     * @return {?}
     */
    function (descriptionPattern) {
        if (!descriptionPattern) {
            return 'empty value not allowed';
        }
        if (/^[a-zA-Z_][a-zA-Z_-]*$/.test(descriptionPattern)) {
            return null; // it is ok
        }
        else {
            return 'description pattern must be an identifier containing only letters, digits, _ or -';
        }
    };
    return NgxTranslateExtractionPattern;
}());
/**
 * Helper class to parse ngx translate extraction pattern
 * and to decide wether a given message matches the pattern.
 */
export { NgxTranslateExtractionPattern };
if (false) {
    /**
     * @type {?}
     * @private
     */
    NgxTranslateExtractionPattern.prototype._matchExplicitId;
    /**
     * @type {?}
     * @private
     */
    NgxTranslateExtractionPattern.prototype._descriptionPatterns;
    /**
     * @type {?}
     * @private
     */
    NgxTranslateExtractionPattern.prototype.extractionPatternString;
}
