// @ngx-i18nsupport/ngx-i18nsupport/esm5/autotranslate/auto-translate-summary-report.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { format } from 'util';
/**
 * A report about a run of Google Translate over all untranslated unit.
 * * Created by martin on 29.06.2017.
 */
var /**
 * A report about a run of Google Translate over all untranslated unit.
 * * Created by martin on 29.06.2017.
 */
AutoTranslateSummaryReport = /** @class */ (function () {
    function AutoTranslateSummaryReport(from, to) {
        this._from = from;
        this._to = to;
        this._total = 0;
        this._ignored = 0;
        this._success = 0;
        this._failed = 0;
    }
    /**
     * Set error if total call failed (e.g. "invalid api key" or "no connection" ...)
     * @param error error
     * @param total total
     */
    /**
     * Set error if total call failed (e.g. "invalid api key" or "no connection" ...)
     * @param {?} error error
     * @param {?} total total
     * @return {?}
     */
    AutoTranslateSummaryReport.prototype.setError = /**
     * Set error if total call failed (e.g. "invalid api key" or "no connection" ...)
     * @param {?} error error
     * @param {?} total total
     * @return {?}
     */
    function (error, total) {
        this._error = error;
        this._total = total;
        this._failed = total;
    };
    /**
     * @return {?}
     */
    AutoTranslateSummaryReport.prototype.error = /**
     * @return {?}
     */
    function () {
        return this._error;
    };
    /**
     * @param {?} ignored
     * @return {?}
     */
    AutoTranslateSummaryReport.prototype.setIgnored = /**
     * @param {?} ignored
     * @return {?}
     */
    function (ignored) {
        this._total += ignored;
        this._ignored = ignored;
    };
    /**
     * Add a single result to the summary.
     * @param tu tu
     * @param result result
     */
    /**
     * Add a single result to the summary.
     * @param {?} tu tu
     * @param {?} result result
     * @return {?}
     */
    AutoTranslateSummaryReport.prototype.addSingleResult = /**
     * Add a single result to the summary.
     * @param {?} tu tu
     * @param {?} result result
     * @return {?}
     */
    function (tu, result) {
        this._total++;
        if (result.success()) {
            this._success++;
        }
        else {
            this._failed++;
        }
    };
    /**
     * Merge another summary into this one.
     * @param anotherSummary anotherSummary
     */
    /**
     * Merge another summary into this one.
     * @param {?} anotherSummary anotherSummary
     * @return {?}
     */
    AutoTranslateSummaryReport.prototype.merge = /**
     * Merge another summary into this one.
     * @param {?} anotherSummary anotherSummary
     * @return {?}
     */
    function (anotherSummary) {
        if (!this._error) {
            this._error = anotherSummary._error;
        }
        this._total += anotherSummary.total();
        this._ignored += anotherSummary.ignored();
        this._success += anotherSummary.success();
        this._failed += anotherSummary.failed();
    };
    /**
     * @return {?}
     */
    AutoTranslateSummaryReport.prototype.total = /**
     * @return {?}
     */
    function () {
        return this._total;
    };
    /**
     * @return {?}
     */
    AutoTranslateSummaryReport.prototype.ignored = /**
     * @return {?}
     */
    function () {
        return this._ignored;
    };
    /**
     * @return {?}
     */
    AutoTranslateSummaryReport.prototype.success = /**
     * @return {?}
     */
    function () {
        return this._success;
    };
    /**
     * @return {?}
     */
    AutoTranslateSummaryReport.prototype.failed = /**
     * @return {?}
     */
    function () {
        return this._failed;
    };
    /**
     * Human readable version of report
     */
    /**
     * Human readable version of report
     * @return {?}
     */
    AutoTranslateSummaryReport.prototype.content = /**
     * Human readable version of report
     * @return {?}
     */
    function () {
        /** @type {?} */
        var result;
        if (this._error) {
            result = format('Auto translation from "%s" to "%s" failed: "%s", failed units: %s', this._from, this._to, this._error, this._failed);
        }
        else {
            result = format('Auto translation from "%s" to "%s", total auto translated units: %s, ignored: %s, succesful: %s, failed: %s', this._from, this._to, this._total, this._ignored, this._success, this._failed);
        }
        return result;
    };
    return AutoTranslateSummaryReport;
}());
/**
 * A report about a run of Google Translate over all untranslated unit.
 * * Created by martin on 29.06.2017.
 */
export { AutoTranslateSummaryReport };
if (false) {
    /**
     * @type {?}
     * @private
     */
    AutoTranslateSummaryReport.prototype._error;
    /**
     * @type {?}
     * @private
     */
    AutoTranslateSummaryReport.prototype._from;
    /**
     * @type {?}
     * @private
     */
    AutoTranslateSummaryReport.prototype._to;
    /**
     * @type {?}
     * @private
     */
    AutoTranslateSummaryReport.prototype._total;
    /**
     * @type {?}
     * @private
     */
    AutoTranslateSummaryReport.prototype._ignored;
    /**
     * @type {?}
     * @private
     */
    AutoTranslateSummaryReport.prototype._success;
    /**
     * @type {?}
     * @private
     */
    AutoTranslateSummaryReport.prototype._failed;
}
