// @ismaestro/ngx-example-library/esm2015/lib/ngx-example-library.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { APP_INITIALIZER, InjectionToken, NgModule } from '@angular/core';
import { NgxExampleLibraryComponent } from './ngx-example-library.component';
import { NgxExampleLibraryService } from './ngx-example-library.service';
/** @type {?} */
export const OPTIONS = new InjectionToken('OPTIONS');
/**
 * @record
 */
export function SampleModuleOptions() { }
/** @type {?} */
SampleModuleOptions.prototype.config;
/**
 * @param {?} options
 * @return {?}
 */
export function initialize(options) {
    console.log('Angular library has started with this options: ', options);
    return function () {
    };
}
export class NgxExampleLibraryModule {
    /**
     * @param {?=} options
     * @return {?}
     */
    static forRoot(options) {
        return {
            ngModule: NgxExampleLibraryModule,
            providers: [
                { provide: OPTIONS, useValue: options },
                {
                    provide: APP_INITIALIZER,
                    useFactory: initialize,
                    deps: [OPTIONS],
                    multi: true
                },
                NgxExampleLibraryService
            ]
        };
    }
}
NgxExampleLibraryModule.decorators = [
    { type: NgModule, args: [{
                imports: [],
                declarations: [
                    NgxExampleLibraryComponent
                ],
                exports: [
                    NgxExampleLibraryComponent
                ]
            },] },
];
