// @ismaestro/ngx-example-library/fesm5/ismaestro-ngx-example-library.js
import { Injectable, Component, APP_INITIALIZER, InjectionToken, NgModule, defineInjectable } from '@angular/core';
import * as moment_ from 'moment';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
var NgxExampleLibraryService = /** @class */ (function () {
    function NgxExampleLibraryService() {
    }
    NgxExampleLibraryService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] },
    ];
    /** @nocollapse */
    NgxExampleLibraryService.ctorParameters = function () { return []; };
    /** @nocollapse */ NgxExampleLibraryService.ngInjectableDef = defineInjectable({ factory: function NgxExampleLibraryService_Factory() { return new NgxExampleLibraryService(); }, token: NgxExampleLibraryService, providedIn: "root" });
    return NgxExampleLibraryService;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/** @type {?} */
var moment = moment_;
var NgxExampleLibraryComponent = /** @class */ (function () {
    function NgxExampleLibraryComponent() {
    }
    /**
     * @return {?}
     */
    NgxExampleLibraryComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.date = moment().format('dddd, D MMMM, YYYY LT');
    };
    NgxExampleLibraryComponent.decorators = [
        { type: Component, args: [{
                    selector: 'lib-ngx-example-library',
                    template: "<span>{{date}}</span>",
                    styles: []
                },] },
    ];
    /** @nocollapse */
    NgxExampleLibraryComponent.ctorParameters = function () { return []; };
    return NgxExampleLibraryComponent;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/** @type {?} */
var OPTIONS = new InjectionToken('OPTIONS');
/**
 * @param {?} options
 * @return {?}
 */
function initialize(options) {
    console.log('Angular library has started with this options: ', options);
    return function () {
    };
}
var NgxExampleLibraryModule = /** @class */ (function () {
    function NgxExampleLibraryModule() {
    }
    /**
     * @param {?=} options
     * @return {?}
     */
    NgxExampleLibraryModule.forRoot = /**
     * @param {?=} options
     * @return {?}
     */
    function (options) {
        return {
            ngModule: NgxExampleLibraryModule,
            providers: [
                { provide: OPTIONS, useValue: options },
                {
                    provide: APP_INITIALIZER,
                    useFactory: initialize,
                    deps: [OPTIONS],
                    multi: true
                },
                NgxExampleLibraryService
            ]
        };
    };
    NgxExampleLibraryModule.decorators = [
        { type: NgModule, args: [{
                    imports: [],
                    declarations: [
                        NgxExampleLibraryComponent
                    ],
                    exports: [
                        NgxExampleLibraryComponent
                    ]
                },] },
    ];
    return NgxExampleLibraryModule;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */

export { NgxExampleLibraryService, NgxExampleLibraryComponent, OPTIONS, initialize, NgxExampleLibraryModule };
