// @ismaestro/ngx-example-library/esm5/lib/ngx-example-library.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component } from '@angular/core';
import * as moment_ from 'moment';
/** @type {?} */
var moment = moment_;
var NgxExampleLibraryComponent = /** @class */ (function () {
    function NgxExampleLibraryComponent() {
    }
    /**
     * @return {?}
     */
    NgxExampleLibraryComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.date = moment().format('dddd, D MMMM, YYYY LT');
    };
    NgxExampleLibraryComponent.decorators = [
        { type: Component, args: [{
                    selector: 'lib-ngx-example-library',
                    template: "<span>{{date}}</span>",
                    styles: []
                },] },
    ];
    /** @nocollapse */
    NgxExampleLibraryComponent.ctorParameters = function () { return []; };
    return NgxExampleLibraryComponent;
}());
export { NgxExampleLibraryComponent };
if (false) {
    /** @type {?} */
    NgxExampleLibraryComponent.prototype.date;
}
