// @ngx-translate/i18n-polyfill/esm5/src/i18n-polyfill.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Inject, Injectable, InjectionToken, LOCALE_ID, MissingTranslationStrategy, Optional, TRANSLATIONS, TRANSLATIONS_FORMAT } from "@angular/core";
import { xliffDigest, xliffLoadToI18n } from "./serializers/xliff";
import { xliff2Digest, xliff2LoadToI18n } from "./serializers/xliff2";
import { xtbDigest, xtbLoadToI18n, xtbMapper } from "./serializers/xtb";
import { HtmlParser, TranslationBundle } from "./parser/html";
import { serializeNodes } from "./serializers/serializer";
/**
 * @record
 */
export function I18nDef() { }
function I18nDef_tsickle_Closure_declarations() {
    /** @type {?} */
    I18nDef.prototype.value;
    /** @type {?|undefined} */
    I18nDef.prototype.id;
    /** @type {?|undefined} */
    I18nDef.prototype.meaning;
    /** @type {?|undefined} */
    I18nDef.prototype.description;
}
export var /** @type {?} */ MISSING_TRANSLATION_STRATEGY = new InjectionToken("MissingTranslationStrategy");
/**
 * A speculative polyfill to use i18n code translations
 */
var I18n = /** @class */ (function () {
    function I18n(format, translations, locale, missingTranslationStrategy) {
        if (missingTranslationStrategy === void 0) { missingTranslationStrategy = MissingTranslationStrategy.Warning; }
        var /** @type {?} */ loadFct;
        var /** @type {?} */ digest;
        var /** @type {?} */ createMapper = function (message) { return null; };
        format = (format || "xlf").toLowerCase();
        switch (format) {
            case "xtb":
                loadFct = xtbLoadToI18n;
                digest = xtbDigest;
                createMapper = xtbMapper;
                break;
            case "xliff2":
            case "xlf2":
                loadFct = xliff2LoadToI18n;
                digest = xliff2Digest;
                break;
            case "xliff":
            case "xlf":
                loadFct = xliffLoadToI18n;
                digest = xliffDigest;
                break;
            default:
                throw new Error("Unknown translations format " + format);
        }
        var /** @type {?} */ htmlParser = new HtmlParser();
        var /** @type {?} */ translationsBundle = TranslationBundle.load(translations, "i18n", digest, createMapper, loadFct, missingTranslationStrategy);
        // todo use interpolation config
        return function (def, params) {
            if (params === void 0) { params = {}; }
            var /** @type {?} */ content = typeof def === "string" ? def : def.value;
            var /** @type {?} */ metadata = {};
            if (typeof def === "object") {
                metadata["id"] = def.id;
                metadata["meaning"] = def.meaning;
                metadata["description"] = def.description;
            }
            var /** @type {?} */ htmlParserResult = htmlParser.parse(content, "", true);
            if (htmlParserResult.errors.length) {
                throw htmlParserResult.errors;
            }
            var /** @type {?} */ mergedNodes = htmlParser.mergeTranslations(htmlParserResult.rootNodes, translationsBundle, params, metadata, ["wrapper"]);
            return serializeNodes(mergedNodes.rootNodes, locale, params).join("");
        };
    }
    I18n.decorators = [
        { type: Injectable },
    ];
    /** @nocollapse */
    I18n.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [TRANSLATIONS_FORMAT,] },] },
        { type: undefined, decorators: [{ type: Inject, args: [TRANSLATIONS,] },] },
        { type: undefined, decorators: [{ type: Inject, args: [LOCALE_ID,] },] },
        { type: MissingTranslationStrategy, decorators: [{ type: Optional }, { type: Inject, args: [MISSING_TRANSLATION_STRATEGY,] },] },
    ]; };
    return I18n;
}());
export { I18n };
function I18n_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    I18n.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    I18n.ctorParameters;
}
