// @ismaestro/ngx-example-library/bundles/ismaestro-ngx-example-library.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('moment')) :
    typeof define === 'function' && define.amd ? define('@ismaestro/ngx-example-library', ['exports', '@angular/core', 'moment'], factory) :
    (factory((global.ismaestro = global.ismaestro || {}, global.ismaestro['ngx-example-library'] = {}),global.ng.core,null));
}(this, (function (exports,i0,moment_) { 'use strict';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    var NgxExampleLibraryService = (function () {
        function NgxExampleLibraryService() {
        }
        NgxExampleLibraryService.decorators = [
            { type: i0.Injectable, args: [{
                        providedIn: 'root'
                    },] },
        ];
        /** @nocollapse */
        NgxExampleLibraryService.ctorParameters = function () { return []; };
        /** @nocollapse */ NgxExampleLibraryService.ngInjectableDef = i0.defineInjectable({ factory: function NgxExampleLibraryService_Factory() { return new NgxExampleLibraryService(); }, token: NgxExampleLibraryService, providedIn: "root" });
        return NgxExampleLibraryService;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    /** @type {?} */
    var moment = moment_;
    var NgxExampleLibraryComponent = (function () {
        function NgxExampleLibraryComponent() {
        }
        /**
         * @return {?}
         */
        NgxExampleLibraryComponent.prototype.ngOnInit = /**
         * @return {?}
         */
            function () {
                this.date = moment().format('dddd, D MMMM, YYYY LT');
            };
        NgxExampleLibraryComponent.decorators = [
            { type: i0.Component, args: [{
                        selector: 'lib-ngx-example-library',
                        template: "<span>{{date}}</span>",
                        styles: []
                    },] },
        ];
        /** @nocollapse */
        NgxExampleLibraryComponent.ctorParameters = function () { return []; };
        return NgxExampleLibraryComponent;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    /** @type {?} */
    var OPTIONS = new i0.InjectionToken('OPTIONS');
    /**
     * @param {?} options
     * @return {?}
     */
    function initialize(options) {
        console.log('Angular library has started with this options: ', options);
        return function () {
        };
    }
    var NgxExampleLibraryModule = (function () {
        function NgxExampleLibraryModule() {
        }
        /**
         * @param {?=} options
         * @return {?}
         */
        NgxExampleLibraryModule.forRoot = /**
         * @param {?=} options
         * @return {?}
         */
            function (options) {
                return {
                    ngModule: NgxExampleLibraryModule,
                    providers: [
                        { provide: OPTIONS, useValue: options },
                        {
                            provide: i0.APP_INITIALIZER,
                            useFactory: initialize,
                            deps: [OPTIONS],
                            multi: true
                        },
                        NgxExampleLibraryService
                    ]
                };
            };
        NgxExampleLibraryModule.decorators = [
            { type: i0.NgModule, args: [{
                        imports: [],
                        declarations: [
                            NgxExampleLibraryComponent
                        ],
                        exports: [
                            NgxExampleLibraryComponent
                        ]
                    },] },
        ];
        return NgxExampleLibraryModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */

    exports.NgxExampleLibraryService = NgxExampleLibraryService;
    exports.NgxExampleLibraryComponent = NgxExampleLibraryComponent;
    exports.OPTIONS = OPTIONS;
    exports.initialize = initialize;
    exports.NgxExampleLibraryModule = NgxExampleLibraryModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
