// @ngx-translate/i18n-polyfill/esm5/src/serializers/serializer.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import * as tslib_1 from "tslib";
import * as html from "../ast/ast";
import * as i18n from "../ast/i18n_ast";
import { getHtmlTagDefinition } from "../ast/html_tags";
import { I18nPluralPipe, I18nSelectPipe, NgLocaleLocalization } from "@angular/common";
import { Parser } from "../ast/parser";
import { getXmlTagDefinition } from "../ast/xml_tags";
/**
 * @record
 */
export function I18nMessagesById() { }
function I18nMessagesById_tsickle_Closure_declarations() {
    /* TODO: handle strange member:
    [msgId: string]: i18n.Node[];
    */
}
/**
 * @record
 */
export function XmlMessagesById() { }
function XmlMessagesById_tsickle_Closure_declarations() {
    /* TODO: handle strange member:
    [id: string]: xml.Node;
    */
}
/**
 * @record
 */
export function IcuContent() { }
function IcuContent_tsickle_Closure_declarations() {
    /** @type {?} */
    IcuContent.prototype.cases;
    /** @type {?} */
    IcuContent.prototype.expression;
    /** @type {?} */
    IcuContent.prototype.type;
}
/**
 * @record
 */
export function IcuContentStr() { }
function IcuContentStr_tsickle_Closure_declarations() {
    /** @type {?} */
    IcuContentStr.prototype.cases;
    /** @type {?} */
    IcuContentStr.prototype.expression;
    /** @type {?} */
    IcuContentStr.prototype.type;
}
/**
 * A `PlaceholderMapper` converts placeholder names from internal to serialized representation and
 * back.
 *
 * It should be used for serialization format that put constraints on the placeholder names.
 * @record
 */
export function PlaceholderMapper() { }
function PlaceholderMapper_tsickle_Closure_declarations() {
    /** @type {?} */
    PlaceholderMapper.prototype.toPublicName;
    /** @type {?} */
    PlaceholderMapper.prototype.toInternalName;
}
/**
 * A simple mapper that take a function to transform an internal name to a public name
 */
var /**
 * A simple mapper that take a function to transform an internal name to a public name
 */
SimplePlaceholderMapper = /** @class */ (function (_super) {
    tslib_1.__extends(SimplePlaceholderMapper, _super);
    // create a mapping from the message
    function SimplePlaceholderMapper(message, mapName) {
        var _this = _super.call(this) || this;
        _this.mapName = mapName;
        _this.internalToPublic = {};
        _this.publicToNextId = {};
        _this.publicToInternal = {};
        message.nodes.forEach(function (node) { return node.visit(_this); });
        return _this;
    }
    /**
     * @param {?} internalName
     * @return {?}
     */
    SimplePlaceholderMapper.prototype.toPublicName = /**
     * @param {?} internalName
     * @return {?}
     */
    function (internalName) {
        return this.internalToPublic.hasOwnProperty(internalName) ? this.internalToPublic[internalName] : null;
    };
    /**
     * @param {?} publicName
     * @return {?}
     */
    SimplePlaceholderMapper.prototype.toInternalName = /**
     * @param {?} publicName
     * @return {?}
     */
    function (publicName) {
        return this.publicToInternal.hasOwnProperty(publicName) ? this.publicToInternal[publicName] : null;
    };
    /**
     * @param {?} text
     * @param {?=} context
     * @return {?}
     */
    SimplePlaceholderMapper.prototype.visitText = /**
     * @param {?} text
     * @param {?=} context
     * @return {?}
     */
    function (text, context) {
        return null;
    };
    /**
     * @param {?} ph
     * @param {?=} context
     * @return {?}
     */
    SimplePlaceholderMapper.prototype.visitTagPlaceholder = /**
     * @param {?} ph
     * @param {?=} context
     * @return {?}
     */
    function (ph, context) {
        this.visitPlaceholderName(ph.startName);
        _super.prototype.visitTagPlaceholder.call(this, ph, context);
        this.visitPlaceholderName(ph.closeName);
    };
    /**
     * @param {?} ph
     * @param {?=} context
     * @return {?}
     */
    SimplePlaceholderMapper.prototype.visitPlaceholder = /**
     * @param {?} ph
     * @param {?=} context
     * @return {?}
     */
    function (ph, context) {
        this.visitPlaceholderName(ph.name);
    };
    /**
     * @param {?} ph
     * @param {?=} context
     * @return {?}
     */
    SimplePlaceholderMapper.prototype.visitIcuPlaceholder = /**
     * @param {?} ph
     * @param {?=} context
     * @return {?}
     */
    function (ph, context) {
        this.visitPlaceholderName(ph.name);
    };
    /**
     * @param {?} internalName
     * @return {?}
     */
    SimplePlaceholderMapper.prototype.visitPlaceholderName = /**
     * @param {?} internalName
     * @return {?}
     */
    function (internalName) {
        if (!internalName || this.internalToPublic.hasOwnProperty(internalName)) {
            return;
        }
        var /** @type {?} */ publicName = this.mapName(internalName);
        if (this.publicToInternal.hasOwnProperty(publicName)) {
            // Create a new XMB when it has already been used
            var /** @type {?} */ nextId = this.publicToNextId[publicName];
            this.publicToNextId[publicName] = nextId + 1;
            publicName = publicName + "_" + nextId;
        }
        else {
            this.publicToNextId[publicName] = 1;
        }
        this.internalToPublic[internalName] = publicName;
        this.publicToInternal[publicName] = internalName;
    };
    return SimplePlaceholderMapper;
}(i18n.RecurseVisitor));
/**
 * A simple mapper that take a function to transform an internal name to a public name
 */
export { SimplePlaceholderMapper };
function SimplePlaceholderMapper_tsickle_Closure_declarations() {
    /** @type {?} */
    SimplePlaceholderMapper.prototype.internalToPublic;
    /** @type {?} */
    SimplePlaceholderMapper.prototype.publicToNextId;
    /** @type {?} */
    SimplePlaceholderMapper.prototype.publicToInternal;
    /** @type {?} */
    SimplePlaceholderMapper.prototype.mapName;
}
var /** @type {?} */ i18nSelectPipe = new I18nSelectPipe();
var SerializerVisitor = /** @class */ (function () {
    function SerializerVisitor(locale, params) {
        this.params = params;
        this.i18nPluralPipe = new I18nPluralPipe(new NgLocaleLocalization(locale));
    }
    /**
     * @param {?} element
     * @param {?} context
     * @return {?}
     */
    SerializerVisitor.prototype.visitElement = /**
     * @param {?} element
     * @param {?} context
     * @return {?}
     */
    function (element, context) {
        if (getHtmlTagDefinition(element.name).isVoid) {
            return "<" + element.name + this.serializeNodes(element.attrs, " ") + "/>";
        }
        return "<" + element.name + this.serializeNodes(element.attrs, " ") + ">" + this.serializeNodes(element.children) + "</" + element.name + ">";
    };
    /**
     * @param {?} attribute
     * @param {?} context
     * @return {?}
     */
    SerializerVisitor.prototype.visitAttribute = /**
     * @param {?} attribute
     * @param {?} context
     * @return {?}
     */
    function (attribute, context) {
        return attribute.name + "=\"" + attribute.value + "\"";
    };
    /**
     * @param {?} text
     * @param {?} context
     * @return {?}
     */
    SerializerVisitor.prototype.visitText = /**
     * @param {?} text
     * @param {?} context
     * @return {?}
     */
    function (text, context) {
        return text.value;
    };
    /**
     * @param {?} comment
     * @param {?} context
     * @return {?}
     */
    SerializerVisitor.prototype.visitComment = /**
     * @param {?} comment
     * @param {?} context
     * @return {?}
     */
    function (comment, context) {
        return "<!--" + comment.value + "-->";
    };
    /**
     * @param {?} expansion
     * @param {?} context
     * @return {?}
     */
    SerializerVisitor.prototype.visitExpansion = /**
     * @param {?} expansion
     * @param {?} context
     * @return {?}
     */
    function (expansion, context) {
        var _this = this;
        var /** @type {?} */ cases = {};
        expansion.cases.forEach(function (c) { return (cases[c.value] = _this.serializeNodes(c.expression)); });
        switch (expansion.type) {
            case "select":
                return i18nSelectPipe.transform(this.params[expansion.switchValue] || "", cases);
            case "plural":
                return this.i18nPluralPipe.transform(this.params[expansion.switchValue], cases);
        }
        throw new Error("Unknown expansion type \"" + expansion.type + "\"");
    };
    /**
     * @param {?} expansionCase
     * @param {?} context
     * @return {?}
     */
    SerializerVisitor.prototype.visitExpansionCase = /**
     * @param {?} expansionCase
     * @param {?} context
     * @return {?}
     */
    function (expansionCase, context) {
        return " " + expansionCase.value + " {" + this.serializeNodes(expansionCase.expression) + "}";
    };
    /**
     * @param {?} nodes
     * @param {?=} join
     * @return {?}
     */
    SerializerVisitor.prototype.serializeNodes = /**
     * @param {?} nodes
     * @param {?=} join
     * @return {?}
     */
    function (nodes, join) {
        var _this = this;
        if (join === void 0) { join = ""; }
        if (nodes.length === 0) {
            return "";
        }
        return join + nodes.map(function (a) { return a.visit(_this, null); }).join(join);
    };
    return SerializerVisitor;
}());
function SerializerVisitor_tsickle_Closure_declarations() {
    /** @type {?} */
    SerializerVisitor.prototype.i18nPluralPipe;
    /** @type {?} */
    SerializerVisitor.prototype.params;
}
/**
 * @param {?} nodes
 * @param {?} locale
 * @param {?} params
 * @return {?}
 */
export function serializeNodes(nodes, locale, params) {
    return nodes.map(function (node) { return node.visit(new SerializerVisitor(locale, params), null); });
}
var HtmlToXmlParser = /** @class */ (function () {
    function HtmlToXmlParser(MESSAGE_TAG) {
        this.MESSAGE_TAG = MESSAGE_TAG;
    }
    /**
     * @param {?} content
     * @return {?}
     */
    HtmlToXmlParser.prototype.parse = /**
     * @param {?} content
     * @return {?}
     */
    function (content) {
        this.xmlMessagesById = {};
        var /** @type {?} */ parser = new Parser(getXmlTagDefinition).parse(content, "", false);
        this.errors = parser.errors;
        html.visitAll(this, parser.rootNodes, null);
        return {
            xmlMessagesById: this.xmlMessagesById,
            errors: this.errors
        };
    };
    /**
     * @param {?} element
     * @param {?} context
     * @return {?}
     */
    HtmlToXmlParser.prototype.visitElement = /**
     * @param {?} element
     * @param {?} context
     * @return {?}
     */
    function (element, context) {
        switch (element.name) {
            case this.MESSAGE_TAG:
                var /** @type {?} */ id = element.attrs.find(function (attr) { return attr.name === "id"; });
                if (id) {
                    this.xmlMessagesById[id.value] = /** @type {?} */ ((/** @type {?} */ (element)));
                }
                break;
            default:
                html.visitAll(this, element.children, null);
        }
    };
    /**
     * @param {?} attribute
     * @param {?} context
     * @return {?}
     */
    HtmlToXmlParser.prototype.visitAttribute = /**
     * @param {?} attribute
     * @param {?} context
     * @return {?}
     */
    function (attribute, context) { };
    /**
     * @param {?} text
     * @param {?} context
     * @return {?}
     */
    HtmlToXmlParser.prototype.visitText = /**
     * @param {?} text
     * @param {?} context
     * @return {?}
     */
    function (text, context) { };
    /**
     * @param {?} comment
     * @param {?} context
     * @return {?}
     */
    HtmlToXmlParser.prototype.visitComment = /**
     * @param {?} comment
     * @param {?} context
     * @return {?}
     */
    function (comment, context) { };
    /**
     * @param {?} expansion
     * @param {?} context
     * @return {?}
     */
    HtmlToXmlParser.prototype.visitExpansion = /**
     * @param {?} expansion
     * @param {?} context
     * @return {?}
     */
    function (expansion, context) { };
    /**
     * @param {?} expansionCase
     * @param {?} context
     * @return {?}
     */
    HtmlToXmlParser.prototype.visitExpansionCase = /**
     * @param {?} expansionCase
     * @param {?} context
     * @return {?}
     */
    function (expansionCase, context) { };
    return HtmlToXmlParser;
}());
export { HtmlToXmlParser };
function HtmlToXmlParser_tsickle_Closure_declarations() {
    /** @type {?} */
    HtmlToXmlParser.prototype.errors;
    /** @type {?} */
    HtmlToXmlParser.prototype.xmlMessagesById;
    /** @type {?} */
    HtmlToXmlParser.prototype.MESSAGE_TAG;
}
