// @ngx-i18nsupport/ngx-i18nsupport/esm5/common/command-output.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
/**
 * Created by martin on 17.02.2017.
 * Very simple class to control the output of a command.
 * Output can be errors, warnings, infos and debug-Outputs.
 * The output can be controlled via 2 flags, quiet and verbose.
 * If quit is enabled only error messages are shown.
 * If verbose is enabled, everything is shown.
 * If both are not enabled (the default) errors, warnings and infos are shown.
 * If not are enabled (strange), we assumed the default.
 */
import chalk from 'chalk';
import { format } from 'util';
/** @enum {number} */
var LogLevel = {
    'ERROR': 0,
    'WARN': 1,
    'INFO': 2,
    'DEBUG': 3,
};
LogLevel[LogLevel['ERROR']] = 'ERROR';
LogLevel[LogLevel['WARN']] = 'WARN';
LogLevel[LogLevel['INFO']] = 'INFO';
LogLevel[LogLevel['DEBUG']] = 'DEBUG';
var CommandOutput = /** @class */ (function () {
    function CommandOutput(stdout) {
        this._quiet = false;
        this._verbose = false;
        if (stdout) {
            this.outputStream = stdout;
        }
        else {
            this.outputStream = process.stdout;
        }
    }
    /**
     * @return {?}
     */
    CommandOutput.prototype.setVerbose = /**
     * @return {?}
     */
    function () {
        this._verbose = true;
    };
    /**
     * @return {?}
     */
    CommandOutput.prototype.setQuiet = /**
     * @return {?}
     */
    function () {
        this._quiet = true;
    };
    /**
     * Test, wether verbose is enabled.
     * @return wether verbose is enabled.
     */
    /**
     * Test, wether verbose is enabled.
     * @return {?} wether verbose is enabled.
     */
    CommandOutput.prototype.verbose = /**
     * Test, wether verbose is enabled.
     * @return {?} wether verbose is enabled.
     */
    function () {
        return this._verbose;
    };
    /**
     * Test, wether quiet is enabled.
     * @return wether quiet is enabled.
     */
    /**
     * Test, wether quiet is enabled.
     * @return {?} wether quiet is enabled.
     */
    CommandOutput.prototype.quiet = /**
     * Test, wether quiet is enabled.
     * @return {?} wether quiet is enabled.
     */
    function () {
        return this._quiet;
    };
    /**
     * @param {?} msg
     * @param {...?} params
     * @return {?}
     */
    CommandOutput.prototype.error = /**
     * @param {?} msg
     * @param {...?} params
     * @return {?}
     */
    function (msg) {
        var params = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            params[_i - 1] = arguments[_i];
        }
        this.log(LogLevel.ERROR, msg, params);
    };
    /**
     * @param {?} msg
     * @param {...?} params
     * @return {?}
     */
    CommandOutput.prototype.warn = /**
     * @param {?} msg
     * @param {...?} params
     * @return {?}
     */
    function (msg) {
        var params = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            params[_i - 1] = arguments[_i];
        }
        this.log(LogLevel.WARN, msg, params);
    };
    /**
     * @param {?} msg
     * @param {...?} params
     * @return {?}
     */
    CommandOutput.prototype.info = /**
     * @param {?} msg
     * @param {...?} params
     * @return {?}
     */
    function (msg) {
        var params = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            params[_i - 1] = arguments[_i];
        }
        this.log(LogLevel.INFO, msg, params);
    };
    /**
     * @param {?} msg
     * @param {...?} params
     * @return {?}
     */
    CommandOutput.prototype.debug = /**
     * @param {?} msg
     * @param {...?} params
     * @return {?}
     */
    function (msg) {
        var params = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            params[_i - 1] = arguments[_i];
        }
        this.log(LogLevel.DEBUG, msg, params);
    };
    /**
     * @private
     * @param {?} level
     * @param {?} msg
     * @param {?} params
     * @return {?}
     */
    CommandOutput.prototype.log = /**
     * @private
     * @param {?} level
     * @param {?} msg
     * @param {?} params
     * @return {?}
     */
    function (level, msg, params) {
        if (!this.isOutputEnabled(level)) {
            return;
        }
        /** @type {?} */
        var coloredMessage;
        switch (level) {
            case LogLevel.ERROR:
                coloredMessage = chalk.red('ERROR: ' + msg);
                break;
            case LogLevel.WARN:
                coloredMessage = chalk.magenta('WARNING: ' + msg);
                break;
            default:
                coloredMessage = chalk.gray('* ' + msg);
                break;
        }
        /** @type {?} */
        var outMsg = format.apply(void 0, tslib_1.__spread([coloredMessage], params));
        this.outputStream.write(outMsg + '\n');
    };
    /**
     * @private
     * @param {?} level
     * @return {?}
     */
    CommandOutput.prototype.isOutputEnabled = /**
     * @private
     * @param {?} level
     * @return {?}
     */
    function (level) {
        /** @type {?} */
        var quietEnabled;
        /** @type {?} */
        var verboseEnabled;
        if (this._quiet && this._verbose) {
            quietEnabled = false;
            verboseEnabled = false;
        }
        else {
            quietEnabled = this._quiet;
            verboseEnabled = this._verbose;
        }
        switch (level) {
            case LogLevel.ERROR:
                return true; // always output errors
            case LogLevel.WARN:
                return (!quietEnabled);
            case LogLevel.INFO:
                return (verboseEnabled && !quietEnabled);
            case LogLevel.DEBUG:
                return verboseEnabled;
            default:
                return true;
        }
    };
    return CommandOutput;
}());
export { CommandOutput };
if (false) {
    /**
     * verbose enables output of everything.
     * @type {?}
     */
    CommandOutput.prototype._verbose;
    /**
     * quiet disables output of everything but errors.
     * @type {?}
     */
    CommandOutput.prototype._quiet;
    /**
     * @type {?}
     * @private
     */
    CommandOutput.prototype.outputStream;
}
